### PR TITLE
Added optimal display mode and frame buffer retrieval

### LINKED ISSUE #22: Add a utility function to enter optimal graphical mode
The optimal mode is the one with the highest resolution (pay no mind to the aspect ratio as of now).

Note the following:
* The color mode must be 32-bit color (RGB and BGR are both OK).
* The mode will be selected and activated using GOP.

Example prototype:
```c++
/**
 * @brief Finds and enters the best possible graphical display mode.
 */
uefi::Status enterOptimalGraphicalDisplayMode(const uefi::SystemTable* systemTable);
```
Link to the latest UEFI specification: https://uefi.org/sites/default/files/resources/UEFI_Spec_2_8_final.pdf

### PR DESCRIPTION
**Pull request by aTOz132**
This closes #22.

### CODE REVIEW

File: bootloader/Utils.cpp [hunk 1]
```diff
@@ -20,3 +20,53 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+
+    return foundBestMode ? gop->setMode(gop, bestMode) : uefi::Status::NOT_FOUND;
+}
+
+uint8* utils::getFrameBuffer(const uefi::SystemTable* systemTable)
+{
+    uefi::protocols::GraphicsOutputProtocol* gop = nullptr;
+    uefi::Status result = systemTable->bootServices->locateProtocol(&uefi::protocols::GRAPHICS_OUTPUT_PROTOCOL_GUID, nullptr,
+                          reinterpret_cast<void**>(&gop));
+    if(result != uefi::Status::SUCCESS)
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Odd spacing... Please add an empty line before this one.

File: bootloader/Utils.h [hunk 2]
```diff
@@ -6,4 +6,12 @@ namespace utils {
      * @brief Finds and enters the best possible text display mode.
      */
     uefi::Status enterOptimalTextMode(const uefi::SystemTable* systemTable);
+    /**
+    * @brief Finds and enters the best possible graphical display mode.
+    */
+    uefi::Status enterOptimalGraphicalDisplayMode(const uefi::SystemTable* systemTable);
+   /**
```
Comment (ExhoAR22) [inline bootloader/Utils.h]: Indent this entire comment by one space.

File: bootloader/Utils.cpp [hunk 3]
```diff
@@ -20,3 +20,53 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+               || info->pixelFormat == uefi::protocols::GraphicsPixelFormat::BGR_RESERVED_8_BIT_PER_COLOR)
+            {
+                foundBestMode = true;
+                bestMode = mode;
+                pixelSum = info->horizontalResolution * info->verticalResolution;
+            }
+        }
+    }
+
+    return foundBestMode ? gop->setMode(gop, bestMode) : uefi::Status::NOT_FOUND;
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: I think it's better to return ```uefi::Status::UNSUPPORTED```.
If you disagree, please explain why :)

File: bootloader/Utils.cpp [hunk 4]
```diff
@@ -20,3 +20,53 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
 
     return systemTable->stdout->setMode(systemTable->stdout, bestMode);
 }
+
+uefi::Status utils::enterOptimalGraphicalDisplayMode(const uefi::SystemTable* systemTable) {
+    uint32 bestMode = 0;
+    bool foundBestMode = false;
+    uint32 pixelSum = 0;
+    uintn  sizeInfo = 0;
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Should be renamed to ```infoSize```.
It's more straight-forward.

Comment (aTOz132) [reply bootloader/Utils.cpp]: Done.

File: bootloader/Utils.cpp (hunk 3 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: I agree, every screen has a "best mode" so if we haven't found one its really unsupported on our end. Will be changed to UNSUPPORTED.

File: bootloader/Utils.cpp (hunk 1 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done.

File: bootloader/Utils.h (hunk 2 above)
Comment (aTOz132) [reply bootloader/Utils.h]: Done.

File: bootloader/Utils.cpp [hunk 5]
```diff
@@ -85,4 +129,4 @@ uint8* utils::readFile(const uefi::SystemTable* systemTable, const char16* path)
     }
 
     return buffer;
-}
+}
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Missing empty line at the end of the file.

File: bootloader/Utils.h [hunk 6]
```diff
@@ -6,19 +6,24 @@ namespace utils {
...
      * @brief Returns the byte count that'll fit in the given number of pages.
      */
     uintn byteCount(uintn pageCount);
-
     /**
      * @brief Reads a file's contents. The returned buffer must be freed using the freePool boot service.
      */
     uint8* readFile(const uefi::SystemTable* systemTable, const char16* path);
-}
+}
```
Comment (ExhoAR22) [inline bootloader/Utils.h]: Missing empty line at the end of the file.

File: bootloader/Utils.cpp (hunk 5 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done.

File: bootloader/Utils.h (hunk 6 above)
Comment (aTOz132) [reply bootloader/Utils.h]: Done.

File: bootloader/Utils.cpp [hunk 7]
```diff
@@ -21,6 +21,50 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+uefi::Status utils::enterOptimalGraphicalDisplayMode(const uefi::SystemTable* systemTable) {
+    uint32 bestMode = 0;
+    bool foundBestMode = false;
+    uint32 pixelSum = 0;
+    uintn  infoSize = 0;
+    uefi::protocols::GraphicsOutputModeInformation* info = nullptr;
+    uefi::protocols::GraphicsOutputProtocol* gop = nullptr;
+    uefi::Status result = systemTable->bootServices->locateProtocol(&uefi::protocols::GRAPHICS_OUTPUT_PROTOCOL_GUID, nullptr,
+                          reinterpret_cast<void**>(&gop));
+    if(result != uefi::Status::SUCCESS) {
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Didn't add a space as you've been told between the ```if``` and the ```(``` :(

File: bootloader/Utils.cpp [hunk 8]
```diff
@@ -21,6 +21,50 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+        return result;
+    }
+    
+    for(uint32 mode = 0; mode < gop->mode->maxMode; mode++) {
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Missing a space between the ```for``` and the ```(``` here as well.

File: bootloader/Utils.cpp [hunk 9]
```diff
@@ -21,6 +21,50 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+        result = gop->queryMode(gop, mode, &infoSize, &info);
+        if (result != uefi::Status::SUCCESS) {
+            return result;
+        }
+
+        if(info->verticalResolution * info->horizontalResolution > pixelSum) {
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Same space problem here.

File: bootloader/Utils.cpp [hunk 10]
```diff
@@ -21,6 +21,50 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+            if(info->pixelFormat == uefi::protocols::GraphicsPixelFormat::RGB_RESERVED_8_BIT_PER_COLOR
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Same space problem here.

File: bootloader/Utils.cpp [hunk 11]
```diff
@@ -21,6 +21,50 @@ uefi::Status utils::enterOptimalTextMode(const uefi::SystemTable* systemTable) {
...
+
+    return foundBestMode ? gop->setMode(gop, bestMode) : uefi::Status::UNSUPPORTED;
+}
+
+uint8* utils::getFrameBuffer(const uefi::SystemTable* systemTable) {
+    uefi::protocols::GraphicsOutputProtocol* gop = nullptr;
+    uefi::Status result = systemTable->bootServices->locateProtocol(&uefi::protocols::GRAPHICS_OUTPUT_PROTOCOL_GUID, nullptr,
+                          reinterpret_cast<void**>(&gop));
+
+    if(result != uefi::Status::SUCCESS) {
```
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: Same space problem here.

File: bootloader/Utils.cpp (hunk 7 above)
Comment (ExhoAR22) [inline bootloader/Utils.cpp]: And please add an empty line between this ```if``` statement and the previous one.

File: bootloader/Utils.h [hunk 12]
```diff
@@ -6,17 +6,22 @@ namespace utils {
      * @brief Finds and enters the best possible text display mode.
      */
     uefi::Status enterOptimalTextMode(const uefi::SystemTable* systemTable);
-
+    /**
+     * @brief Finds and enters the best possible graphical display mode.
+     */
+    uefi::Status enterOptimalGraphicalDisplayMode(const uefi::SystemTable* systemTable);
+    /**
```
Comment (ExhoAR22) [inline bootloader/Utils.h]: The way this should be formatted is the following:
```c++
namespace ... {
    namespace ... {
        /**
         * @brief ...
         */
        uefi::Status ...(...);
        
        /**
         * @brief ...
         */
        uefi::Status ...(...);
    }
}
```

Sorry if my ```CONTRIBUTING.md``` example was not clear enough, I'll revamp it soon.

File: bootloader/Utils.cpp (hunk 7 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.cpp (hunk 8 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.cpp (hunk 9 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.cpp (hunk 10 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.cpp (hunk 11 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.cpp (hunk 7 above)
Comment (aTOz132) [reply bootloader/Utils.cpp]: Done for good now.

File: bootloader/Utils.h (hunk 12 above)
Comment (aTOz132) [reply bootloader/Utils.h]: Done for good now.

**Pull request merged**